Source/interactive_control.cpp

```diff
@@ -1,593 +1,620 @@
+        }
+
+        if (success && json.has_field(RPC_DISABLED))
+        {
+            m_disabled = json[RPC_DISABLED].as_bool();
+        }
+        else
+        {
+            if (overwrite)
+            {
+                errorString = L"Trying to construct or update a button, no disabled field";
+                success = false;
+            }
+        }
+
+        if (success && json.has_field(RPC_CONTROL_BUTTON_PROGRESS))
+        {
+            m_progress = (float)json[RPC_CONTROL_BUTTON_PROGRESS].as_number().to_double();
+        }
+    }
+    catch (std::exception e)
+    {
+        LOGS_ERROR << "Failed to parse json in interactive_button_control::update()";
+    }
+
+    return success;
+}
+
+void
+interactive_button_control::clear_state()
+{
+    m_buttonStateBymixerId.clear();
+    m_buttonCount->clear();
+}
+
+interactive_button_control::interactive_button_control()
+{
+    m_interactivityManager = interactivity_manager::get_singleton_instance();
+    m_disabled = true;
+    m_type = interactive_control_type::button;
+    m_buttonCount = std::shared_ptr<interactive_button_count>(new interactive_button_count());
+}
+
+interactive_button_control::interactive_button_control(
+    _In_ string_t parentSceneId,
+    _In_ string_t controlId,
+    _In_ bool disabled,
+    _In_ float progress,
+    _In_ std::chrono::milliseconds cooldownDeadline,
+    _In_ string_t buttonText,
+    _In_ uint32_t sparkCost
+)
+{
+    m_interactivityManager = interactivity_manager::get_singleton_instance();
+    m_parentScene = std::move(parentSceneId);
+    m_type = interactive_control_type::button;
+    m_controlId = std::move(controlId);
+    m_disabled = disabled;
+    m_progress = progress;
+    m_buttonText = std::move(buttonText);
+    m_cooldownDeadline = std::move(cooldownDeadline);
+    m_sparkCost = sparkCost;
+    m_buttonCount = std::shared_ptr<interactive_button_count>(new interactive_button_count());
+}
+
+//
+// Joystick control
+//
+  
+double
+interactive_joystick_control::x() const
+{
+    return m_x;
+}
+
+double
+interactive_joystick_control::x(_In_ uint32_t mixerId)
+{
+    double x = 0;
+    if (m_joystickStateBymixerId[mixerId] != nullptr)
+    {
+        x = m_joystickStateBymixerId[mixerId]->x();
+    }
+    return x;
+}
+
+double
+interactive_joystick_control::y() const
+{
+    return m_y;
+}
+
+double
+interactive_joystick_control::y(_In_ uint32_t mixerId)
+{
+    double y = 0;
+    if (m_joystickStateBymixerId[mixerId] != nullptr)
+    {
+        y = m_joystickStateBymixerId[mixerId]->y();
+    }
+    return y;
+}
+
+void
+interactive_joystick_control::clear_state()
+{
+    // We do not clear the joystick state. The reason is because (1) we will get a 0, 0 joystick
+    // event if the joystick returns to rest. The other reason is if the user is holding the joytick
+    // at a certain value, let's say (1, 1), then we will not continue to recieve input events, however
+    // we need to report (1, 1) as the joystick value.

```

Comment:
yeah makes sense - maybe it should be removed from the base class, if we can get away with that (no dependencies instantiate this class directly)

---
In reply to: [125791774](https://github.com/mixer/interactive-cpp/pull/17#discussion_r125791774) [](ancestors = 125791774)